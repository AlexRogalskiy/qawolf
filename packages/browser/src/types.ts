declare module "puppeteer/lib/USKeyboardLayout" {
  export interface KeyDefinition {
    code: string;
    key: string;
    keyCode?: number;
    location?: number;
    shiftKey?: string;
    shiftKeyCode?: number;
  }

  export interface KeyDefinitionsType {
    [key: string]: KeyDefinition;
  }

  const map: KeyDefinitionsType;
  export default map;
}
