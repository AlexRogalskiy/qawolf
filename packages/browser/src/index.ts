export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
