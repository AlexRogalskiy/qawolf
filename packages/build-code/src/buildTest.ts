import { Workflow } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { resolve } from "path";
import { formatStep } from "./formatStep";

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const buildTest = (workflow: Workflow) => {
  const test = testTemplate({
    name: workflow.name,
    steps: workflow.steps.map(step => formatStep(step)),
    url: workflow.url
  });

  return test;
};
