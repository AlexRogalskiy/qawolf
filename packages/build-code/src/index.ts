export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
