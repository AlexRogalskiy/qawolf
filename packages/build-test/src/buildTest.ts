import { BrowserStep, Job, Action } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { resolve } from "path";

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const formatIt = (step: BrowserStep): string => {
  if (step.action === "scroll") {
    return `can scroll ${step.scrollDirection!}`;
  }

  const stepValue = step.value ? ` into` : "";

  const targetTagName = `${
    step.target.tagName === "a" ? "link" : step.target.tagName || ""
  }`;

  const label = step.target.labels ? step.target.labels[0] : "";

  const targetName =
    label ||
    step.target.name ||
    step.target.placeholder ||
    step.target.textContent ||
    step.target.id ||
    "";
  const truncatedTargetName = targetName.trim().substring(0, 20); // ensure not too long

  return `can ${step.action}${stepValue} ${truncatedTargetName} ${targetTagName}`;
};

export const formatMethod = (action: Action, index: number): string => {
  const stepParam = `steps[${index}]`;

  if (action === "click") {
    return `click(${stepParam})`;
  }

  if (action === "input") {
    return `input(${stepParam}, values[${index}])`;
  }

  if (action === "scroll") {
    return `scroll(${stepParam})`;
  }

  throw new Error(`Invalid step action ${action}`);
};

export const formatStep = (step: BrowserStep) => {
  return {
    it: formatIt(step),
    method: formatMethod(step.action, step.index)
  };
};

export const buildTest = (job: Job) => {
  const test = testTemplate({
    name: job.name,
    steps: job.steps.map(step => formatStep(step))
  });

  return test;
};
