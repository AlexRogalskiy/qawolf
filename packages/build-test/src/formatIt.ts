import { Step } from "@qawolf/types";

export const formatTarget = (step: Step): string => {
  // link/input/select
  const tagName = `${
    step.target.tagName === "a" ? "link" : step.target.tagName || ""
  }`;

  const label = step.target.labels ? step.target.labels.join(" ") : "";

  // ex. "departure date"
  let description =
    label ||
    step.target.name ||
    step.target.placeholder ||
    step.target.innerText ||
    step.target.id ||
    "";

  description = description.trim();

  if (description.length > 40) {
    description = `${description.substring(0, 40)}...`;
  }

  if (description.length) return `"${description}" ${tagName}`;

  return tagName;
};

export const formatIt = (step: Step): string => {
  if (step.action === "scroll") {
    return `can scroll`;
  }

  const target = formatTarget(step);

  if (step.action === "click") {
    return `can click ${target}`;
  }

  //   if (step.action === "select") {
  //     return `can input into ${target}`;
  //   }

  if (step.action === "type") {
    if (!step.value) {
      return `can clear ${target}`;
    }

    const value = step.value as string;
    if (value.indexOf("↓Enter") === 0) {
      return `can Enter`;
    }

    if (value.indexOf("↓Tab") === 0) {
      return `can Tab`;
    }

    return `can type into ${target}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
