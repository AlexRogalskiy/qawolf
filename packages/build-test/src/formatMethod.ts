import { Action } from "@qawolf/types";

export const formatMethod = (action: Action, index: number): string => {
  const stepParam = `steps[${index}]`;

  if (action === "click") {
    return `click(${stepParam})`;
  }

  if (action === "type") {
    return `type(${stepParam}, values[${index}])`;
  }

  if (action === "scroll") {
    return `scroll(${stepParam}, values[${index}])`;
  }

  throw new Error(`Invalid step action ${action}`);
};
