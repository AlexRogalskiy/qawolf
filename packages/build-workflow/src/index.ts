// export * from "./buildClickSteps";
// export * from "./buildScrollSteps";
// export * from "./buildSelectSteps";
// export * from "./buildSteps";
// export * from "./buildTypeSteps";
// export * from "./buildWorkflow";
