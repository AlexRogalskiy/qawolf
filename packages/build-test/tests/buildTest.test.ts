import { buildTest, formatStep } from "../src/buildTest";

// import directly since fixtures are note exported from @qawolf/build-job
import { loginJob } from "../../build-job/fixtures/loginJob";
import { BrowserStep } from "@qawolf/types";

describe("buildTest", () => {
  it("builds a test from a job", async () => {
    const testString = buildTest(loginJob);
    expect(testString).toMatchSnapshot();
  });
});

describe("formatStep", () => {
  it("formats click on link step", () => {
    const step: BrowserStep = {
      action: "click",
      index: 0,
      target: {
        tagName: "a",
        textContent: "contact"
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats click on submit input step", () => {
    const step: BrowserStep = {
      action: "click",
      index: 1,
      target: {
        inputType: "submit",
        tagName: "input",
        textContent: "sign in"
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats scroll down action", () => {
    const step: BrowserStep = {
      action: "scroll",
      index: 0,
      scrollDirection: "down",
      target: { xpath: "scroll" }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats scroll up action", () => {
    const step: BrowserStep = {
      action: "scroll",
      index: 0,
      scrollDirection: "up",
      target: { xpath: "scroll" }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats type into text input", () => {
    const step: BrowserStep = {
      action: "input",
      index: 0,
      target: {
        id: "input1",
        inputType: "text",
        labels: ["username"],
        name: "user",
        placeholder: "Jane Doe",
        tagName: "input"
      },
      value: "spirit"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats type into password input", () => {
    const step: BrowserStep = {
      action: "input",
      index: 10,
      target: {
        id: "input2",
        inputType: "password",
        placeholder: "secret",
        tagName: "input"
      },
      value: "supersecret"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("shortens target name if needed", () => {
    const step: BrowserStep = {
      action: "click",
      index: 0,
      target: {
        inputType: "submit",
        tagName: "input",
        textContent: `sign in${"x".repeat(200)}`
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });
});
