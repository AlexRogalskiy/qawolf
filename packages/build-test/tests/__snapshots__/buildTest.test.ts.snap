// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTest builds a test from a job 1`] = `
"describe('login', () => {
  it('can scroll down', async () => {
    await scroll(steps[0]};
  });

  it('can click form authentication link', async () => {
    await click(steps[1]);
  });

  it('can input into username input', async () => {
    await input(steps[2], values[2]);
  });

  it('can click login button', async () => {
    await click(steps[3]);
  });

  it('can input into username input', async () => {
    await input(steps[4], values[4]);
  });

  it('can input into password input', async () => {
    await input(steps[5], values[5]);
  });

  it('can click login i', async () => {
    await click(steps[6]);
  });
});"
`;

exports[`formatStep formats click on link step 1`] = `
Object {
  "it": "can click contact link",
  "method": "click(steps[0])",
}
`;

exports[`formatStep formats click on submit input step 1`] = `
Object {
  "it": "can click sign in input",
  "method": "click(steps[1])",
}
`;

exports[`formatStep formats scroll down action 1`] = `
Object {
  "it": "can scroll down",
  "method": "scroll(steps[0]}",
}
`;

exports[`formatStep formats scroll up action 1`] = `
Object {
  "it": "can scroll up",
  "method": "scroll(steps[0]}",
}
`;

exports[`formatStep formats type into password input 1`] = `
Object {
  "it": "can input into secret input",
  "method": "input(steps[10], values[10])",
}
`;

exports[`formatStep formats type into text input 1`] = `
Object {
  "it": "can input into username input",
  "method": "input(steps[0], values[0])",
}
`;

exports[`formatStep shortens target name if needed 1`] = `
Object {
  "it": "can click sign inxxxxxxxxxxxxx input",
  "method": "click(steps[0])",
}
`;
